### PR TITLE
Adding elasticache_facts module

### PR DESCRIPTION
**Pull request by magnetik**
Module to get facts from AWS elasticache.

Exemple result :

   <127.0.0.1> REMOTE_MODULE elasticache_facts instance_name=test region=eu-west-1
changed: [54.194.130.20] => {"changed": true, "instance": {"configuration_endpoint": {"address": "YYYYY.euw1.cache.amazonaws.com", "port": 11211}, "engine": "memcached", "id": "test", "instance_type": "cache.t1.micro", "node_endpoints": [{"address": "XXXXX.cache.amazonaws.com", "id": "0001", "port": 11211, "status": "available"}], "status": "available"}, "item": ""}

### CODE REVIEW

**Comment by jctanner** [comment]
@magnetik you have duplicated some of the code in https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/ec2.py but otherwise it looks good.

**Comment by mpdehaan** [comment]
More review pending but a few small things...

Please use the dict constructor for Python dictionaries

```
  d = {
 +      'id': data['CacheClusterId'],
 +      'status': data['CacheClusterStatus'],
+      'instance_type': data['CacheNodeType'],
+      'engine': data['Engine'],
+      'node_endpoints': []
+    }
```

Should be

```
 d = dict(
    id = data['CacheClusterId'],
    ...
 )
```

Also looks like some indent issues across the board, we standardize on 4 spaces, so please update this to be consistent throughout.

Thanks!

**Comment by mpdehaan** [comment]
@jctanner yeah we should definitely also reuse common code from that module.

**Comment by magnetik** [comment]
Both of your remarks should be fixed. Thanks for the quick answer!

**Comment by magnetik** [comment]
Is everything okay now?

**Comment by jimi-c** [comment]
@magnetik yes I believe you addressed the issues raised. We will be getting this in for further review and testing soon. Thanks!

**Comment by pas256** [comment]
Shouldn't `instance_name` be `cache_name` or `cache_cluster_id`?

Code to the abstraction AWS is providing, not the underlying technology.

File: library/cloud/elasticache_facts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        id = data['CacheClusterId'],
+        status = data['CacheClusterStatus'],
+        instance_type = data['CacheNodeType'],
+        engine = data['Engine'],
+        node_endpoints = []
+    )
+
+    if d['status'] == 'available':
+        d['configuration_endpoint'] = dict(
+            address = data['ConfigurationEndpoint']['Address'],
```
Comment (Anislav) [inline library/cloud/elasticache_facts:93]: I am running this module against Redis cache cluster and it works for me when the following if clause is added:

```
if data['ConfigurationEndpoint']:
     d['configuration_endpoint'] = dict(
          address=data['ConfigurationEndpoint']['Address'],
          port=data['ConfigurationEndpoint']['Port']
     )
```

This is due to the following reason: at this time Auto Discovery is only available for cache clusters running the Memcached engine as it is stated in the documentation: http://docs.aws.amazon.com/AmazonElastiCache/latest/UserGuide/AutoDiscovery.html

File: library/cloud/elasticache_facts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        nodeData = dict(
+            id = node['CacheNodeId'],
+            status = node['CacheNodeStatus']
+        )
+        if nodeData['status'] == 'available':
+            nodeData['address'] = node['Endpoint']['Address']
+            nodeData['port'] = node['Endpoint']['Port']
+        d['node_endpoints'].append(nodeData)
+
+    module.exit_json(changed=True, instance=d)
```
Comment (Anislav) [inline library/cloud/elasticache_facts:107]: I am new to ansible but I understand the documentation (http://docs.ansible.com/developing_modules.html#module-provided-facts) in a way that a module which provides facts should return a dictionary that contains a ansible_facts key.
For example ec2_facts module works in the way described above so the users like me would expect elasticache_facts module to work in the same way as the others.

**Comment by mpdehaan** [comment]
My thoughts about this one now are that the inventory script could really be returning the list about all the enumerated caches in the various regions, keeping people from having to explicitly call a module to get them. 

I think we should update the inventory script to provide this data in hash form by cluster/etc so that users could just index it without having to call the extra module and go through the dance with register.

Sorry for the delay in getting to this -- we've had quite a backlog to power through!

**Comment by magnetik** [comment]
No problem, thanks for the input :)

**Pull request closed without merging**